joint_trajectory_controller/test/test_trajectory_actions.cpp

```diff
@@ -0,0 +1,595 @@
+
+  GoalOptions opt;
+  opt.goal_response_callback =
+    std::bind(&TestTrajectoryActions::common_goal_response, this, _1);
+  opt.result_callback =
+    std::bind(&TestTrajectoryActions::common_result_response, this, _1);
+  opt.feedback_callback = nullptr;
+
+  // send goal
+  {
+    std::vector<JointTrajectoryPoint> points;
+    JointTrajectoryPoint point;
+    point.time_from_start.set__sec(0);  // start asap
+    point.positions.resize(joint_names.size());
+
+    point.positions[0] = 1.0;
+    point.positions[1] = 2.0;
+    point.positions[2] = 3.0;
+    points.push_back(point);
+
+    sendActionGoal(points, 1, opt);
+  }
+  controller_hw_thread.join();
+
+  EXPECT_EQ(true, common_goal_accepted);
+  EXPECT_EQ(rclcpp_action::ResultCode::SUCCEEDED, common_resultcode);
+  EXPECT_EQ(
+    control_msgs::action::FollowJointTrajectory_Result::SUCCESSFUL,
+    common_action_result_code);
+
+  EXPECT_NEAR(1.0, test_robot->pos1, common_threshold);
+  EXPECT_NEAR(2.0, test_robot->pos2, common_threshold);
+  EXPECT_NEAR(3.0, test_robot->pos3, common_threshold);
+
+  // start again
+  common_goal_accepted = false;
+  common_resultcode = rclcpp_action::ResultCode::UNKNOWN;
+  controller_hw_thread = std::thread(thread_func);
+
+  // add feedback
+  bool feedback_recv = false;
+  opt.feedback_callback = [&](
+    rclcpp_action::ClientGoalHandle<FollowJointTrajectoryMsg>::SharedPtr,
+    const std::shared_ptr<const FollowJointTrajectoryMsg::Feedback> feedback)
+    {
+      (void)feedback;
+      feedback_recv = true;
+    };
+
+  // send goal with multiple points
+  {
+    std::vector<JointTrajectoryPoint> points;
+    JointTrajectoryPoint point1;
+    point1.time_from_start.set__sec(0);
+    point1.time_from_start.set__nanosec(200000000);  // 0.2 seconds
+    point1.positions.resize(joint_names.size());
+
+    point1.positions[0] = 4.0;
+    point1.positions[1] = 5.0;
+    point1.positions[2] = 6.0;
+    points.push_back(point1);
+
+    JointTrajectoryPoint point2;
+    point2.time_from_start.set__sec(0);
+    point2.time_from_start.set__nanosec(300000000);  // 0.3 seconds
+    point2.positions.resize(joint_names.size());
+
+    point2.positions[0] = 7.0;
+    point2.positions[1] = 8.0;
+    point2.positions[2] = 9.0;
+    points.push_back(point2);
+
+    sendActionGoal(points, 1, opt);
+  }
+  controller_hw_thread.join();
+
+  EXPECT_EQ(true, feedback_recv);
+  EXPECT_EQ(true, common_goal_accepted);
+  EXPECT_EQ(rclcpp_action::ResultCode::SUCCEEDED, common_resultcode);
+  EXPECT_EQ(
+    control_msgs::action::FollowJointTrajectory_Result::SUCCESSFUL,
+    common_action_result_code);
+
+  EXPECT_NEAR(7.0, test_robot->pos1, common_threshold);
+  EXPECT_NEAR(8.0, test_robot->pos2, common_threshold);
+  EXPECT_NEAR(9.0, test_robot->pos3, common_threshold);
+
+  executor.cancel();
+}
+
+TEST_F(TestTrajectoryActions, test_state_tolerances_fail) {
+  auto traj_controller = std::make_shared<joint_trajectory_controller::JointTrajectoryController>();
```

Comment:
shouldn't this entire block be part of the test-case setup?

```
  auto traj_controller = std::make_shared<joint_trajectory_controller::JointTrajectoryController>();
  auto ret = traj_controller->init(test_robot, controller_name);
  if (ret != controller_interface::CONTROLLER_INTERFACE_RET_SUCCESS) {
    FAIL();
  }

  auto traj_lifecycle_node = traj_controller->get_lifecycle_node();
  rclcpp::Parameter joint_parameters("joints", joint_names);
  traj_lifecycle_node->set_parameter(joint_parameters);

  rclcpp::Parameter operation_mode_parameters("write_op_modes", op_mode);
  traj_lifecycle_node->set_parameter(operation_mode_parameters);
```

Reply:
abstract code to test setup a2a02dc